core/types/block.go

```diff
@@ -220,10 +221,20 @@ type extblock struct {
 // The values of TxHash, UncleHash, ReceiptHash and Bloom in header
 // are ignored and set to values derived from the given txs, uncles
 // and receipts.
-func NewBlock(header *Header, txs []*Transaction, uncles []*Header, receipts []*Receipt, hasher TrieHasher) *Block {
-	b := &Block{header: CopyHeader(header)}
+func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher) *Block {
```

Comment:
Given that a Body is fairly small containing only slices, and you also have this strange check below to handle nil,it might be simpler to go with the struct.

Reply:
Unrelated: it's a bit confusing when you should use NewBlock and when NewBlockWithHeader.WithBody. Perhaps we could consider tweaking the names of one or the other a bit to make it clearer that they do different things and serve different purposes. Otherwise IMO it's asking for it.

Reply:
Since we already handle `Body` as a pointer value in other places, I think making it a struct in `NewBlock` is going to be a bit annoying. This is seen mostly in implementations of `consensus.Engine` where `FinalizeAndAssemble` takes a pointer `Body`.

If you think it is worth turning all uses of `Body` over to struct values, I can do it. It just seems like it might actually be better to keep the pointer value given this.

Similarly with `NewBlock` and `WithBody` interfaces matching, if they need to match, they should probably both be `*Body`.

cc @fjl since I think it was his thought to convert to struct val.

Reply:
Also agree on it being confusing when to use `NewBlock` vs. `NewBlockWithHeader.WithBody`. I think the answer is you use `NewBlock` when you need to compute the header accumulators with the body values and `NewBlockWithHeader.WithBody` when you need to recreate the block which has been disassembled (written to disk, received over the wire, etc).

Naming isn't perfect, but struggling to come up with something markedly better.

Reply:
I still think it's a bit non-obvious, but it might suffice to update some method-docs.

Reply:
The difference with `NewBlock` is that it would deep copy all inputs. It's meant to be used in tests, and that's also why it has a short name. Across production code, very few places actually *create blocks*, so we can live with a longer function name and we can also avoid some of the copying.